src/Deserialization.cpp

```diff
@@ -0,0 +1,1358 @@
+    memcpy(hl_buffer.data(), buffer->data()->data(), buffer->data()->size());
+    return hl_buffer;
+}
+
+std::map<std::string, FunctionPtr> Deserializer::deserialize_wrapper_refs(const flatbuffers::Vector<flatbuffers::Offset<Serialize::WrapperRef>> *wrappers) {
+    assert(wrappers != nullptr);
+    std::map<std::string, FunctionPtr> result;
+    for (const auto &wrapper : *wrappers) {
+        auto name = deserialize_string(wrapper->func_name());
+        int32_t func_index = wrapper->func_index();
+        FunctionPtr func_ptr;
+        if (auto it = this->reverse_function_mappings.find(func_index); it != this->reverse_function_mappings.end() && func_index != -1) {
+            func_ptr = it->second;
+        }
+        result[name] = func_ptr;
+    }
+    return result;
+}
+
+void Deserializer::build_reverse_function_mappings(const std::vector<Function> &functions) {
+    if (!this->reverse_function_mappings.empty()) {
+        this->reverse_function_mappings.clear();
+    }
+    int cnt = 0;
+    for (const auto &f : functions) {
+        this->reverse_function_mappings[cnt++] = f.get_contents();
+    }
+}
+
+Pipeline Deserializer::deserialize(const std::string &filename) {
+    std::ifstream in(filename, std::ios::binary | std::ios::in);
+    if (!in) {
+        user_assert(false) << "failed to open file " << filename << "\n";
+        return Pipeline();
+    }
+    in.seekg(0, std::ios::end);
+    int size = in.tellg();
+    in.seekg(0, std::ios::beg);
+    std::vector<char> data(size);
+    in.read(data.data(), size);
+    in.close();
+
+    const auto *pipeline_obj = Serialize::GetPipeline(data.data());
+    std::vector<std::string> func_names_in_order;
+    for (const auto &func_name : *pipeline_obj->func_names_in_order()) {
+        func_names_in_order.push_back(deserialize_string(func_name));
+    }
+
+    // We use the first realized function to build the group and all other functions below to this same group
+    std::vector<Function> functions;
+    functions.reserve(func_names_in_order.size());
+    if (!func_names_in_order.empty()) {
+        functions.push_back(Function(func_names_in_order[0]));
+        for (size_t i = 1; i < func_names_in_order.size(); ++i) {
+            functions.push_back(functions[0].new_function_in_same_group(func_names_in_order[i]));
+        }
+    }
+    build_reverse_function_mappings(functions);
+
+    // Buffers need to be deserialized first as Parameters may reference them
+    std::vector<Buffer<>> buffers;
+    buffers.reserve(pipeline_obj->buffers()->size());
+    for (const auto &buffer : *pipeline_obj->buffers()) {
+        buffers.push_back(deserialize_buffer(buffer));
+    }
+    for (const auto &buffer : buffers) {
+        if (buffers_in_pipeline.find(buffer.name()) == buffers_in_pipeline.end()) {
+            buffers_in_pipeline[buffer.name()] = buffer;
+        } else {
+            user_assert(false) << "duplicate buffer " << buffer.name() << " in pipeline\n";
```

Comment:
here and elsewhere: why not just do

```
            user_assert(buffers_in_pipeline.count(buffer.name()) > 0) << "duplicate buffer " << buffer.name() << " in pipeline\n";
            buffers_in_pipeline[buffer.name()] = buffer;
```

(The `else` can be elided since user_assert will never return if the condition is false, and `count()` is terser than find/end since we just care about presence. Sadly, `contains()` is C++20 and we haven't yet migrated.)